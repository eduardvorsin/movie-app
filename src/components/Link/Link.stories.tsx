import type { Meta, StoryObj } from '@storybook/react';
import Link from './Link';


const meta: Meta<typeof Link> = {
	title: 'UI/Link',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for the Link componentz'
		},
		id: {
			description: 'Unique identifier',
		},
		children: {
			description: 'Component content',
		},
		href: {
			description: 'The url to which the link will be clicked',
		},
		onClick: {
			description: 'The callback that is called when the link is clicked',
		},
		onFocus: {
			description: 'The callback that is called when a link is hit',
		},
		onBlur: {
			description: 'A callback that is called when the focus is lost from the link',
		},
		isDisabled: {
			description: 'Makes the link inactive for interaction',
		},
		target: {
			description: 'Allows you to set the way a new page will open when a link is clicked',
		},
		isExternal: {
			description: 'Determines whether the link is internal or external',
		},
		testId: {
			description: 'Id for testing the component',
		}
	},
	component: Link,
	parameters: {
		docs: {
			description: {
				component: 'Links transport users to a different location and typically manifest within or immediately after a sentence.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof Link>;

export const Default: Story = {
	args: {
		children: 'Default',
		href: '/',
	},
	parameters: {
		docs: {
			description: {
				story: 'Default state of the component',
			},
		},
	},
};

export const Disabled: Story = {
	args: {
		children: 'Disabled',
		isDisabled: true,
		href: '/',
	},
	parameters: {
		docs: {
			description: {
				story: 'The inactive state of a link is useful when you need to temporarily restrict the user\'s ability to interact with it.',
			},
		},
	},
};

export const External: Story = {
	args: {
		children: 'External',
		isExternal: true,
		href: 'https://www.google.com/',
	},
	parameters: {
		docs: {
			description: {
				story: 'Allows you to define a link to an external resource. Also implements additional security mechanisms for this',
			},
		},
	},
};

export const EventCallbacks: Story = {
	argTypes: {
		onClick: {
			action: 'Clicked',
		},
		onFocus: {
			action: 'Focused',
		},
		onBlur: {
			action: 'Blured',
		}
	},
	args: {
		children: 'click, focus, blur events',
		href: '/',
	},
	parameters: {
		docs: {
			description: {
				story: 'Callback functions can be supplied to onClick, onBlur, and onFocus events',
			},
		},
	},
};