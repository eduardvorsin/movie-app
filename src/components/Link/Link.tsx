import { AnchorHTMLAttributes, CSSProperties, FocusEventHandler, MouseEventHandler, ReactNode } from 'react';
import NextLink from 'next/link';

type Props = {
	children: ReactNode,
	href: string,
	id?: string,
	className?: string,
	onClick?: MouseEventHandler<HTMLAnchorElement>,
	onFocus?: FocusEventHandler<HTMLAnchorElement>,
	onBlur?: FocusEventHandler<HTMLAnchorElement>,
	isDisabled?: boolean,
	target?: AnchorHTMLAttributes<HTMLButtonElement>['target'],
	isExternal?: boolean,
	testId?: string,
	style?: CSSProperties,
};

export default function Link({
	id,
	href,
	className,
	children,
	onClick,
	onFocus,
	onBlur,
	isExternal = false,
	testId,
	isDisabled = false,
	style,
	target,
}: Props) {
	const linkClasses = [
		'underline text-blue-500 hover:text-blue-600 hover:no-underline active:text-blue-700',
		isDisabled ? 'opacity-disabled cursor-not-allowed' : '',
		className,
	].join(' ');

	if (!isExternal) {
		return (
			<NextLink
				id={id}
				className={linkClasses}
				href={href}
				onClick={onClick}
				onFocus={onFocus}
				onBlur={onBlur}
				role={!href ? 'link' : undefined}
				aria-disabled={!href}
				data-testid={testId}
				style={style}
			>
				{children}
			</NextLink>
		);
	}

	return (
		<a
			id={id}
			className={`${linkClasses}`}
			href={href}
			rel='noopener noreferrer'
			onClick={onClick}
			aria-disabled={!href}
			target={target}
			role={!href ? 'link' : undefined}
			data-testid={testId}
			style={style}
		>
			{children}
		</a>
	);
};
