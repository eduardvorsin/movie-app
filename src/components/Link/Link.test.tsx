import { render, screen } from "@testing-library/react";
import Link from "./Link";
import userEvent from "@testing-library/user-event";

describe('Link tests', () => {
	it('is rendered correctly', () => {
		render(<Link href='/'>test link</Link>);

		expect(screen.getByRole<HTMLAnchorElement>('link')).toBeInTheDocument();
	});
	it('when you click on the link, the mock function is triggered', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();
		render(<Link href='/' onClick={mockFn}>test link</Link>);

		await user.click(screen.getByRole<HTMLAnchorElement>('link'));

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('when the link receives the focus, the mock function is triggered', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();
		render(<Link href='/' onFocus={mockFn}>test link</Link>);

		await user.tab();

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('when the focus is lost, the mock function is triggered', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();
		render(<Link href='/' onFocus={mockFn}>test link</Link>);

		await user.tab();
		await user.click(document.body);

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('if isDisabled is true then the link becomes inactive', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();
		render(
			<Link
				href='/'
				onFocus={mockFn}
				isDisabled
			>
				test link
			</Link>
		);

		await user.click(screen.getByRole<HTMLAnchorElement>('link'));

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('is a basic snapshot', () => {
		render(<Link href='/'>test link</Link>);

		expect(screen.getByRole<HTMLAnchorElement>('link')).toMatchSnapshot();
	});

	it('is a snapshot with isDisabled set to true', () => {
		render(<Link href='/' isDisabled>test link</Link>);

		expect(screen.getByRole<HTMLAnchorElement>('link')).toMatchSnapshot();
	});

	it('is a snapshot with isExternal equal to true', () => {
		render(<Link href='/' isExternal>test link</Link>);

		expect(screen.getByRole<HTMLAnchorElement>('link')).toMatchSnapshot();
	});
});