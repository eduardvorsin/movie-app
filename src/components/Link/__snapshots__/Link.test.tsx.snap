// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link tests is a basic snapshot 1`] = `
<a
  aria-disabled="false"
  class="no-underline  hover:underline transition-colors duration-150 text-blue-800 hover:text-blue-900 active:text-blue-1000 dark:text-blue-400 dark:hover:text-blue-300 dark:active:text-blue-200  "
  href="/"
>
  test link
</a>
`;

exports[`Link tests is a snapshot with isDisabled set to true 1`] = `
<a
  aria-disabled="false"
  class="no-underline  hover:underline transition-colors duration-150 text-blue-800 hover:text-blue-900 active:text-blue-1000 dark:text-blue-400 dark:hover:text-blue-300 dark:active:text-blue-200 opacity-disabled cursor-not-allowed "
  href="/"
>
  test link
</a>
`;

exports[`Link tests is a snapshot with isExternal equal to true 1`] = `
<a
  aria-disabled="false"
  class="no-underline  hover:underline transition-colors duration-150 text-blue-800 hover:text-blue-900 active:text-blue-1000 dark:text-blue-400 dark:hover:text-blue-300 dark:active:text-blue-200  "
  href="/"
  rel="noopener noreferrer"
>
  test link
</a>
`;
